File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        # Bottom up Dynamic Programming
        last_day = days[-1]
        dp, i = [0] * (last_day + 1), 0

        for day in range(1, last_day + 1):
            if day < days[i]:
                dp[day] = dp[day - 1]
            else:
                i += 1
                dp[day] = min(dp[day - 1] + costs[0],\
                            dp[max(0, day - 7)] + costs[1],\
                            dp[max(0, day - 30)] + costs[2])

        return dp[last_day]

File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.cpp
class Solution {
private:
    unordered_set<int> isTravelNeeded;

    int solve(vector<int>& days, vector<int>& costs, vector<int>& dp, int currDay){
        // Base case: If we've iterated over travel days, return 0.
        if (currDay > days.back())
            return 0;

        // If already calculated, return from here with the stored answer.
        if (dp[currDay] != -1)
            return dp[currDay];

        // If we don't need to travel on this day, move on to next day.
        if (!isTravelNeeded.count(currDay))
            return solve(days, costs, dp, currDay + 1);
        
        int oneDay = costs[0] + solve(days, costs, dp, currDay + 1);
        int sevenDays = costs[1] + solve(days, costs, dp, currDay + 7);
        int thirtyDays = costs[2] + solve(days, costs, dp, currDay + 30);

        // Store the cost with the minimum of the three options.
        return dp[currDay] = min(oneDay, min(sevenDays, thirtyDays));
    }

public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        // The last day on which we need to travel.
        int lastDay = days.back();
        vector<int> dp(lastDay + 1, -1);

        // Mark the days on which we need to travel.
        for (int day : days)
            isTravelNeeded.insert(day);

        return solve(days, costs, dp, 1);
    }
};